docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+As quick overview let's consider stream below:
+
+```json
+{
+  "stream": {
+   "job": "loki-dev/ingester",
+   "cluster": "us-central1",
+   "namespace" : "loki-dev",
+  },
+  "values": [
+    [
+      "1584114643856675277",
+      "level=info ts=2020-03-13T15:50:43.856555 caller=metrics.go"
+    ],
+    [
+      "1584114643856675278",
+      "level=warn ts=2020-03-13T15:50:43.856556 caller=metrics.go"
+    ]
+  ]
+}
+```
+
+As you can see the log is using the popular logfmt so we could use the `| logfmt` parsers to transform those entries into the fields entries like below:
+
+```json
+{
+  "stream": {
+   "job": "loki-dev/ingester",
+   "cluster": "us-central1",
+   "namespace" : "loki-dev",
+  },
+  "values": [
+    [
+      "1584114643856675277",
+      {
+          "level":"info",
+          "ts": "2020-03-13T15:50:43.856555",
+          "caller": "metrics.go"
+      }
+    ],
+   [
+      "1584114643856675278",
+      {
+          "level":"warn",
+          "ts": "2020-03-13T15:50:43.856556",
+          "caller": "metrics.go"
+      }
+    ]
+  ]
+}
+```
+
+Field names are declared by each parsers, some parser will allow implicit fields extraction for __brevity__ like the `json` and `logfmt` parsers.
+
+Since field name can conflicts with labels during aggregation we will support for each parser the ability to rename a field. However in case of conflict fields will take precedence over labels.
+
+> Implicit fields extraction could means trouble in term of performance. However in most cases we will be able to infer all used fields from the full query and so limit the extraction per line.
+
+To cover a wide range of structured and unstructured logs, we will start with those four parsers: `glob`, `logfmt`, `json` and `regexp`.
+
+### Glob
+
+The glob parser (as described in previous design) `{job="prod/nginx"} | glob "*<client_ip> - - [ * ] \"* *<path> "` could be used to extract fields from nginx logs.
+
+The log entry below:
+
+```log
+127.0.0.1 - - [19/Jun/2012:09:16:22 +0100] "GET /GO.jpg HTTP/1.1" 499 0 "http://domain.com/htm_data/7/1206/758536.html" "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; .NET CLR 1.1.4322; .NET CLR 2.0.50727; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; SE 2.X MetaSr 1.0)"
+```
+
+would be transformed into those fields:
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "path": "/GO.jpg"
+}
+```
+
+This parser use [glob pattern matching](https://en.wikipedia.org/wiki/Glob_(programming)) and extract named matches(*\<name>) into fields.
+If the line doesn't match the glob it will be filtered out.
+
+### Json
+
+The json parser allows to parse a json log line and select fields to extract. If the line is not a valid json document the line will be skipped.
+
+We could support those notations:
+
+```logql
+{job="prod/app"} | json .route client_ip=.client.ip
+{job="prod/app"} | json .route .client.ip as route, client_ip
```

Comment:
It may be less beneficial to expose both `=` and `as` for assigning keys to values. Unless I'm missing something, I think it'd be better to choose one.

Reply:
Yeah I'm asking for feedback which one is better ?

Also do you prefer ` | json .route, .client` or `|json .route .client` ?

Reply:
`| json`